### PR TITLE
Add `--diff` option `ruff format`

### LINKED ISSUE #7231: Formatter: Add an option to show formatting changes when running `--check`
There are two options:

1. Add a new `--diff` option to `ruff format --check` that prints the formatting changes similar to Black
2. `--check` uses Ruff's diagnostic system and shows diffs for `format: human` but not for `format: compact` (which would enable diffs by default)

I prefer option 2. 

Open questions
* Should this be integrated into our diagnostic system? 
* Should it print the entire diff or collapse very large code frames
* Which diffing library to use. We ran into exponential back-tracking cases in our `cargo dev` script for large files / large changes

CC @zanieb regarding CLI design

### PR DESCRIPTION
**Pull request by konstin**
**Summary** `ruff format --diff` is similar to `ruff format --check`, but we don't only error with the list of file that would be formatted, but also show a diff between the unformatted input and the formatted output.

```console
$ ruff format --diff scratch.py scratch.pyi scratch.ipynb
warning: `ruff format` is not yet stable, and subject to change in future versions.
--- scratch.ipynb
+++ scratch.ipynb
@@ -1,3 +1,4 @@
 import numpy
-maths = (numpy.arange(100)**2).sum()
-stats= numpy.asarray([1,2,3,4]).median()
+
+maths = (numpy.arange(100) ** 2).sum()
+stats = numpy.asarray([1, 2, 3, 4]).median()
--- scratch.py
+++ scratch.py
@@ -1,3 +1,3 @@
 x = 1
-y=2
+y = 2
 z = 3
2 files would be reformatted, 1 file left unchanged
```

With `--diff`, the summary message gets printed to stderr to allow e.g. `ruff format --diff . > format.patch`.

At the moment, jupyter notebooks are formatted as code diffs, while everything else is a real diff that could be applied. This means that the diffs containing jupyter notebooks are not real diffs and can't be applied. We could change this to json diffs, but they are hard to read. We could also split the diff option into a human diff option, where we deviate from the machine readable diff constraints, and a proper machine readable, appliable diff output that you can pipe into other tools.

To make the tests work, the results (and errors, if any) are sorted before printing them. Previously, the print order was random, i.e. two identical runs could have different output.

Open question: Should this go into the markdown docs? Or will this be subsumed by the integration of the formatter into `ruff check`?

**Test plan** Fixtures for the change and no change cases, including a jupyter notebook and for file input and stdin.

Fixes #7231

### CODE REVIEW

File: crates/ruff_cli/src/commands/format.rs [hunk 1]
```diff
@@ -155,47 +193,50 @@ fn format_path(
...
                 formatted
                     .write(&mut writer)
                     .map_err(|err| FormatCommandError::Write(Some(path.to_path_buf()), err))?;
+                Ok(FormatResult::Formatted)
             }
-            Ok(FormatResult::Formatted)
-        }
+            FormatMode::Check => Ok(FormatResult::Formatted),
+            FormatMode::Diff => Ok(FormatResult::Diff {
+                path: path.to_path_buf(),
```
Comment (charliermarsh) [inline crates/ruff_cli/src/commands/format.rs:211]: Can we make this unowned, by storing `&Path`? The lifetimes should work out, as far as I can tell.

File: crates/ruff_cli/src/commands/format.rs [hunk 2]
```diff
@@ -347,6 +399,23 @@ impl Display for FormatSummary<'_> {
                     formatted += 1;
                 }
                 FormatResult::Unchanged => unchanged += 1,
+                FormatResult::Diff {
+                    path,
+                    unformatted,
+                    formatted: formatted_source,
+                } => {
+                    // Diff
```
Comment (charliermarsh) [inline crates/ruff_cli/src/commands/format.rs:407]: Nit: can probably remove this comment, it doesn't add much on top of the match guard itself (`FormatResult::Diff`).

File: crates/ruff_cli/src/commands/format.rs [hunk 3]
```diff
@@ -347,6 +399,23 @@ impl Display for FormatSummary<'_> {
                     formatted += 1;
                 }
                 FormatResult::Unchanged => unchanged += 1,
+                FormatResult::Diff {
+                    path,
+                    unformatted,
+                    formatted: formatted_source,
```
Comment (charliermarsh) [inline crates/ruff_cli/src/commands/format.rs:405]: Why rename here, but leave `unformatted` as-is? Can we either leave them both, or rename them both, for consistency?

File: crates/ruff_cli/src/commands/format.rs [hunk 4]
```diff
@@ -155,47 +193,50 @@ fn format_path(
...
 }
 
 #[derive(Debug)]
 pub(crate) enum FormattedSource {
     /// The source was formatted, and the [`SourceKind`] contains the transformed source code.
-    Formatted(SourceKind),
+    Formatted {
+        unformatted: SourceKind,
+        formatted: SourceKind,
+    },
```
Comment (charliermarsh) [inline crates/ruff_cli/src/commands/format.rs:226]: This is making me wonder if `format_source` should take `&SourceKind`, and we should change `Unchanged(SourceKind)` to just `Unchanged`, and remove `unformatted` from the `Formatted` variant. It strikes me as inflexible that the method takes ownership of `SourceKind` but only acts on references and then returns the owned `SourceKind` in all cases. But, this is minor -- if it doesn't work for whatever reason, or you disagree, it's fine.

**Review by charliermarsh — APPROVED** [review]

**Comment by charliermarsh** [comment]
> Should this go into the markdown docs?

I think it's fine to leave this as-is for now -- I will fold into the documentation when I revisit it next week.

File: crates/ruff_cli/src/commands/format.rs (hunk 4 above)
Comment (MichaReiser) [reply crates/ruff_cli/src/commands/format.rs:226]: I guess the challenge with this is that we need to move `SourceKind` out of the linter, maybe into `ruff_source_file`? I recommend investigating this as a separate PR.

File: crates/ruff_cli/src/commands/format.rs [hunk 5]
```diff
@@ -347,6 +399,23 @@ impl Display for FormatSummary<'_> {
                     formatted += 1;
                 }
                 FormatResult::Unchanged => unchanged += 1,
+                FormatResult::Diff {
+                    path,
+                    unformatted,
+                    formatted: formatted_source,
+                } => {
+                    // Diff
+                    let mut writer = stdout().lock();
```
Comment (MichaReiser) [inline crates/ruff_cli/src/commands/format.rs:408]: Should `stdout` be locked outside of the loop?

File: crates/ruff_cli/src/commands/format.rs [hunk 6]
```diff
@@ -347,6 +399,23 @@ impl Display for FormatSummary<'_> {
...
+                    let text_diff = TextDiff::from_lines(
+                        unformatted.source_code(),
+                        formatted_source.source_code(),
+                    );
+                    let mut unified_diff = text_diff.unified_diff();
+                    unified_diff.header(&fs::relativize_path(path), &fs::relativize_path(path));
+                    unified_diff.to_writer(&mut writer).unwrap();
+
+                    formatted += 1;
+                }
```
Comment (MichaReiser) [inline crates/ruff_cli/src/commands/format.rs:418]: Nit: It feels a bit strange that the diff is written as part of the `FormatSummary`'s `Display` implementation (the diff isn't part of the summary). 

I would prefer moving the diff printing out of the summary implementation and handle it explicitly when the `FormatMode` is `Diff` (also allows to locking `stdout` exactly once)

**Review by MichaReiser — APPROVED** [review]

File: crates/ruff_cli/src/commands/format.rs (hunk 4 above)
Comment (konstin) [reply crates/ruff_cli/src/commands/format.rs:226]: Thought so too but didn't want to create so much churn, changed it now

File: crates/ruff_cli/src/commands/format.rs (hunk 1 above)
Comment (konstin) [reply crates/ruff_cli/src/commands/format.rs:211]: I missed that we already have the path on the outer struct that we can use

**Comment by konstin** [comment]
> I guess the challenge with this is that we need to move SourceKind out of the linter, maybe into ruff_source_file? I recommend investigating this as a separate PR.

Did work without it, but i'm fine with moving it anyway

File: crates/ruff_cli/src/commands/format.rs (hunk 6 above)
Comment (konstin) [reply crates/ruff_cli/src/commands/format.rs]: I went with renaming `FormatSummary` to `FormatResults` and giving it two different write methods. We unfortunately lose the fmt impl but we get proper separation

**Pull request merged**